File: packages/sdk/src/tests/unit/snapshotMigration0005.test.ts
import { SnapshotSchema } from '@towns-protocol/proto'
import { bin_toHexString } from '@towns-protocol/dlog'
import { create } from '@bufbuild/protobuf'
import { snapshotMigration0005 } from '../../migrations/snapshotMigration0005'
import { streamIdToBytes } from '../../id'
import { makeUniqueSpaceStreamId } from '../testUtils'

describe('snapshotMigration0005', () => {
    test('run migration with common session IDs and empty solicitations', () => {
        // Create a test session ID that will appear frequently
        const commonSessionIdBytes = new Uint8Array(
            'common-session-id-5'.split('').map((c) => c.charCodeAt(0)),
        )
        const commonSessionId = bin_toHexString(commonSessionIdBytes)
        const spaceId = makeUniqueSpaceStreamId()
        const spaceIdBytes = streamIdToBytes(spaceId)

        // Create a snapshot with multiple members having the same session ID
        const snap = create(SnapshotSchema, {
            content: {
                case: 'spaceContent',
                value: {
                    inception: {
                        streamId: spaceIdBytes,
                    },
                },
            },
            members: {
                joined: [
                    {
                        solicitations: [
                            { sessionIds: [commonSessionId, 'unique1_a'] },
                            { sessionIds: [commonSessionId] }, // Will become empty
                        ],
                        username: {
                            data: {
                                sessionId: commonSessionId,
                                algorithm: 'test-algo',
                            },
                        },
                        displayName: {
                            data: {
                                sessionIdBytes: commonSessionIdBytes,
                                algorithm: 'test-algo',
                            },
                        },
                    },
                    {
                        solicitations: [{ sessionIds: [commonSessionId, 'unique2'] }],
                        username: {
                            data: {
                                sessionIdBytes: commonSessionIdBytes,
                                algorithm: 'test-algo',
                            },
                        },
                    },
                    {
                        solicitations: [{ sessionIds: [commonSessionId, 'unique3'] }],
                    },
                    {
                        solicitations: [{ sessionIds: ['unique4'] }],
                    },
                ],
            },
        })

        const result = snapshotMigration0005(snap, true)

        for (const member of result.members!.joined) {
            for (const solicitation of member.solicitations) {
                expect(solicitation.sessionIds).not.toContain(commonSessionId)
            }
            // All solicitations should be non-empty
            for (const solicitation of member.solicitations) {
                expect(solicitation.sessionIds.length).toBeGreaterThan(0)
            }
            // Usernames and display names with common session IDs should be cleared
            if (member.username?.data) {
                expect(member.username.data.sessionId).not.toBe(commonSessionId)
                if (member.username.data.sessionIdBytes?.length) {
                    expect(bin_toHexString(member.username.data.sessionIdBytes)).not.toBe(
                        commonSessionId,
                    )
                }
            }
            if (member.displayName?.data) {
                expect(member.displayName.data.sessionId).not.toBe(commonSessionId)
                if (member.displayName.data.sessionIdBytes?.length) {
                    expect(bin_toHexString(member.displayName.data.sessionIdBytes)).not.toBe(
                        commonSessionId,
                    )
                }
            }
        }
    })

    test('run migration with empty members', () => {
        const snap = create(SnapshotSchema, {})
        const result = snapshotMigration0005(snap, true)
        expect(result).toEqual(snap)
    })

    test('run migration with no common session IDs', () => {
        const snap = create(SnapshotSchema, {
            members: {
                joined: [
                    {
                        solicitations: [{ sessionIds: ['unique1', 'unique2'] }],
                        username: {
                            data: {
                                sessionId: 'unique3',
                                algorithm: 'test-algo',
                            },
                        },
                    },
                    {
                        solicitations: [{ sessionIds: ['unique4', 'unique5'] }],
                    },
                ],
            },
        })

        const result = snapshotMigration0005(snap, true)
        // Verify the snapshot remains unchanged when no common session IDs are found
        expect(result).toEqual(snap)
    })
})

File: core/node/events/migrations/snapshot_migration_0005_test.go
package migrations

import (
	"encoding/hex"
	"testing"

	"github.com/stretchr/testify/require"

	"google.golang.org/protobuf/proto"

	. "github.com/towns-protocol/towns/core/node/protocol"
	"github.com/towns-protocol/towns/core/node/shared"
)

func TestSnapshotMigration0005(t *testing.T) {
	t.Run("with common session IDs and empty solicitations", func(t *testing.T) {
		commonSessionIdBytes := []byte("common-session-id-5")
		commonSessionId := hex.EncodeToString(commonSessionIdBytes)
		spaceId, err := shared.MakeSpaceId()
		require.NoError(t, err)

		snap := &Snapshot{
			Content: &Snapshot_SpaceContent{
				SpaceContent: &SpacePayload_Snapshot{
					Inception: &SpacePayload_Inception{
						StreamId: spaceId[:],
					},
				},
			},
			Members: &MemberPayload_Snapshot{
				Joined: []*MemberPayload_Snapshot_Member{
					{
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{commonSessionId, "unique1_a"}},
							{SessionIds: []string{commonSessionId}}, // Will become empty
						},
						Username: &WrappedEncryptedData{
							Data: &EncryptedData{
								SessionId: commonSessionId,
								Algorithm: "test-algo",
							},
						},
						DisplayName: &WrappedEncryptedData{
							Data: &EncryptedData{
								SessionIdBytes: commonSessionIdBytes,
								Algorithm:      "test-algo",
							},
						},
					},
					{
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{commonSessionId, "unique2"}},
						},
						Username: &WrappedEncryptedData{
							Data: &EncryptedData{
								SessionIdBytes: commonSessionIdBytes,
								Algorithm:      "test-algo",
							},
						},
					},
					{
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{commonSessionId, "unique3"}},
						},
					},
					{
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{"unique4"}},
						},
					},
				},
			},
		}

		snapshot_migration_0005_(snap, true)

		for _, member := range snap.GetMembers().GetJoined() {
			for _, solicitation := range member.GetSolicitations() {
				require.NotContains(t, solicitation.GetSessionIds(), commonSessionId)
			}
			// All solicitations should be non-empty
			for _, solicitation := range member.GetSolicitations() {
				require.NotEmpty(t, solicitation.GetSessionIds())
			}
			// Usernames and display names with common session IDs should be cleared
			if member.GetUsername() != nil {
				require.NotEqual(t, member.GetUsername().GetData().GetSessionId(), commonSessionId)
				if len(member.GetUsername().GetData().GetSessionIdBytes()) > 0 {
					require.NotEqual(t, string(member.GetUsername().GetData().GetSessionIdBytes()), commonSessionId)
				}
			}
			if member.GetDisplayName() != nil {
				require.NotEqual(t, member.GetDisplayName().GetData().GetSessionId(), commonSessionId)
				if len(member.GetDisplayName().GetData().GetSessionIdBytes()) > 0 {
					require.NotEqual(t, string(member.GetDisplayName().GetData().GetSessionIdBytes()), commonSessionId)
				}
			}
		}
	})

	t.Run("with empty members", func(t *testing.T) {
		snap := &Snapshot{}
		snapshot_migration_0005_(snap, true)
		require.NotNil(t, snap)
	})

	t.Run("with no common session IDs", func(t *testing.T) {
		snap := &Snapshot{
			Members: &MemberPayload_Snapshot{
				Joined: []*MemberPayload_Snapshot_Member{
					{
						UserAddress: []byte("member1"),
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{"unique1", "unique2"}},
						},
						Username: &WrappedEncryptedData{
							Data: &EncryptedData{
								SessionId: "unique3",
								Algorithm: "test-algo",
							},
						},
					},
					{
						UserAddress: []byte("member2"),
						Solicitations: []*MemberPayload_KeySolicitation{
							{SessionIds: []string{"unique4", "unique5"}},
						},
					},
				},
			},
		}
		snapCopy, err := proto.Marshal(snap)
		require.NoError(t, err)
		snapshot_migration_0005_(snap, true)
		resultBytes, err := proto.Marshal(snap)
		require.NoError(t, err)
		require.Equal(t, snapCopy, resultBytes)
	})
}
